### ISSUE TITLE
anyOf, allOf taking varargs: 0 args should not be allowed

### ISSUE THREAD

**Issue by alb-i986** [hamcrest/JavaHamcrest#147]
Matchers taking varargs are error-prone.
One may have a test like this:

    [..]
    assertThat(subject, allOf());

I think this can't be anything but a mistake, therefore I'd like it to be reported as an error to the user.

As a solution, I think the best feedback would be to trigger a compilation error, by changing the signature:

    public static <T> org.hamcrest.core.AnyOf<T> anyOf(Matcher<? super T> first, Matcher<? super T>... matchers)

instead of 

    public static <T> org.hamcrest.core.AnyOf<T> anyOf(Matcher<? super T>... matchers)

Possibly other matchers in `Matchers` accepting varargs are affected as well.

**Comment by pawelfanduel** [comment]
Yeah. Methods `Matchers.contains` and `Macthers.containsInAnyOrder` are affected in the worst way. If your matchers generic type are not a perfect for it it treats them as `Object` and accepts as parameters of `Matchers.contains(E... items)` without issuing error. Takes a bit to notice that generics are the cause of it.
